src/mysqlsh_helpers.py

```diff
@@ -0,0 +1,292 @@
+            cluster_name: cluster name
+            root_password: password for the 'root' user
+            server_config_user: user name for the server config user
+            server_config_password: password for the server config user
+            cluster_admin_user: user name for the cluster admin user
+            cluster_admin_password: password for the cluster admin user
+            container: workload container object
+        """
+        self.instance_address = instance_address
+        self.cluster_name = cluster_name
+        self.root_password = root_password
+        self.server_config_user = server_config_user
+        self.server_config_password = server_config_password
+        self.cluster_admin_user = cluster_admin_user
+        self.cluster_admin_password = cluster_admin_password
+        self.container = container
+
+    def configure_mysql_users(self) -> None:
+        """Configure the MySQL users for the instance.
+
+        Creates base `root@%` and `<server_config>@%` users with the
+        appropriate privileges, and reconfigure `root@localhost` user password.
+
+        Raises MySQLConfigureMySQLUsersError if the user creation fails.
+        """
+        # SYSTEM_USER and SUPER privileges to revoke from the root users
+        # Reference: https://dev.mysql.com/doc/refman/8.0/en/privileges-provided.html#priv_super
+        privileges_to_revoke = (
+            "SYSTEM_USER",
+            "SYSTEM_VARIABLES_ADMIN",
+            "SUPER",
+            "REPLICATION_SLAVE_ADMIN",
+            "GROUP_REPLICATION_ADMIN",
+            "BINLOG_ADMIN",
+            "SET_USER_ID",
+            "ENCRYPTION_KEY_ADMIN",
+            "VERSION_TOKEN_ADMIN",
+            "CONNECTION_ADMIN",
+        )
+
+        # it's not needed to grant privileges to the root user, as it's already
+        # granted by the entrypoint script provided by the container
+        configure_users_commands = (
+            "UPDATE mysql.user SET authentication_string=null WHERE User='root' and Host='%';",
+            f"ALTER USER 'root'@'%' IDENTIFIED BY '{self.root_password}';",
+            f"CREATE USER '{self.server_config_user}'@'%' IDENTIFIED BY '{self.server_config_password}';",
+            f"GRANT ALL ON *.* TO '{self.server_config_user}'@'%' WITH GRANT OPTION;",
+            "UPDATE mysql.user SET authentication_string=null WHERE User='root' and Host='localhost';",
+            f"ALTER USER 'root'@'localhost' IDENTIFIED BY '{self.root_password}';",
+            f"REVOKE {', '.join(privileges_to_revoke)} ON *.* FROM 'root'@'%';",
+            f"REVOKE {', '.join(privileges_to_revoke)} ON *.* FROM 'root'@'localhost';",
+            "FLUSH PRIVILEGES;",
+        )
+
+        try:
+            logger.debug("Configuring MySQL users")
+            self._run_mysqlcli_script(" ".join(configure_users_commands))
+        except ExecError as e:
+            logger.error("Exited with code %d. Stderr:", e.exit_code)
+            if e.stderr:
```

Comment:
It looks like that `stderr` will always be None if [wait](https://ops.readthedocs.io/en/latest/index.html?highlight=execerror#ops.pebble.ExecError.stderr) is called instead of `wait_output` in the `_run_mysqlcli_script` function.

Reply:
good catch. It helped me catching a wrong exception being raised too. Done on 3081db5